server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java

```diff
@@ -83,92 +123,150 @@ public void beforeRefresh() throws IOException {
     /**
      * Upload new segment files created as part of the last refresh to the remote segment store.
      * This method also uploads remote_segments_metadata file which contains metadata of each segment file uploaded.
+     *
      * @param didRefresh true if the refresh opened a new reference
      */
     @Override
     public void afterRefresh(boolean didRefresh) {
-        synchronized (this) {
-            try {
-                if (indexShard.getReplicationTracker().isPrimaryMode()) {
-                    if (this.primaryTerm != indexShard.getOperationPrimaryTerm()) {
-                        this.primaryTerm = indexShard.getOperationPrimaryTerm();
-                        this.remoteDirectory.init();
+        try {
+            indexShard.getThreadPool().executor(ThreadPool.Names.REMOTE_REFRESH).submit(() -> syncSegments(false)).get();
+        } catch (InterruptedException | ExecutionException e) {
+            logger.info("Exception occurred while scheduling syncSegments", e);
+        }
+    }
+
+    private synchronized void syncSegments(boolean isRetry) {
```

Comment:
consider this - normal refresh triggered upload failed, it scheduled a retry scheduled at n seconds from the failure time. Now meanwhile another normal refresh triggered upload failed, it will try to trigger another retry scheduled at 2*n seconds from then but the first retry is yet to happen. If we do not have the semaphore (permits check), multiple retry can get scheduled while not even a single retry has really happened. Also, semaphore helps schedule deterministic exponential back off retries by ensuring one retry can be scheduled at a time. when the retry fails, it will schedule next refresh at 2*n seconds from then. hope this makes sense.